File: DSA-Python/5-array/11-max-consecutive-ones.py
# Given a binary array nums, return the maximum number of consecutive 1's in the array.



# brute force - O(n), O(1)
def maximum_consecutive_ones_1(nums:list[int]) -> int:
    cnt, max_ones = 0, 0
    
    for x in nums:
        if (x != 1):
            cnt = 0
        else:
            cnt += 1
            max_ones = max(max_ones, cnt)
            
    return max_ones
    
    

# sliding window - O(n), O(1)
def maximum_consecutive_ones_2(nums:list[int]) -> int:
    left, n, max_ones = 0, len(nums), 0
    
    for right in range(n):
        if nums[right] == 0:
            left = right + 1
        else:
            max_ones = max(max_ones, right-left+1)
            
    return max_ones



if __name__ == "__main__": 
    loop = int(input())

    for _ in range(loop):
        nums = list(map(int, input().split()))
        
        res = maximum_consecutive_ones_1(nums)
        
        res = maximum_consecutive_ones_2(nums)
        
        print(res)
        
        
'''
5
1 1 0 1 1 1
1 0 1 1 0 1
1

0
'''

File: DSA-C++/6-array/11-max-consecutive-ones.cpp
#include <bits/stdc++.h>

using namespace std;
// Given an array that contains only 1 and 0 return the count of maximum consecutive ones in the array..


// brute force - O(n), O(1)
int max_consecutive_ones(vector <int> &nums) {
    int n = nums.size(), max_ones = 0, cnt = 0;

    for (int i=0; i<n; ++i) {
        if (nums[i] == 0 && cnt > max_ones) {
            max_ones = cnt;
            cnt = 0;
        }
        else {
            cnt++;
        }
    }

    if (cnt > max_ones) {
        max_ones = cnt;
    }

    return max_ones;
}

int main() {
    vector <int> nums = {1, 1, 0, 1, 1, 1};
    
    cout << max_consecutive_ones(nums) << endl;

    return 0;
}